### PR TITLE
Controller: make harvesting stronger/more resilient

### PR DESCRIPTION
**Pull request by sergiogiro**

### CODE REVIEW

File: services/controller/src/plz/controller/containers.py [hunk 1]
```diff
@@ -78,7 +78,7 @@ def rm(self, execution_id: str):
     def get_state(self, execution_id: str) -> Optional[ContainerState]:
```
Comment (SamirTalwar) [inline services/controller/src/plz/controller/containers.py]: No longer `Optional`.

File: services/controller/src/plz/controller/results/results_base.py [hunk 2]
```diff
@@ -15,6 +18,20 @@ def publish(self,
...
+                raise CouldNotGetOutputException(
+                    f'Couldn\'t read the results for {execution_id}')
+            else:
+                # Make sure the logs are non-empty
+                try:
+                    next(results.logs())
+                except StopIteration:
+                    raise CouldNotGetOutputException(
+                        f'Suspicious empty logs for {execution_id}')
+        log.debug(f'Logs are available for {execution_id}')
```
Comment (SamirTalwar) [inline services/controller/src/plz/controller/results/results_base.py]: Don't do this in the abstract class, please.

File: services/controller/src/plz/controller/instances/docker.py [hunk 3]
```diff
@@ -131,6 +138,8 @@ def release(self,
     def _do_release(self, results_storage: ResultsStorage):
         self.stop_execution()
         self._publish_results(results_storage)
+        # Check that we could collect the logs before destroying the container
+        results_storage.check_logs_available(self.execution_id)
```
Comment (SamirTalwar) [inline services/controller/src/plz/controller/instances/docker.py]: Check that the `finished_file` has been written, not the logs.

In the interface:

```python
    @abstractmethod
    def is_finished(self, execution_id: str):
        pass
```

In the `LocalResultsStorage` implementation:

```python
    def is_finished(self, execution_id: str):
        paths = Paths(self.directory, execution_id)
        return os.path.exists(paths.finished_file)
```

File: services/controller/src/plz/controller/results/results_base.py (hunk 2 above)
Comment (SamirTalwar) [reply services/controller/src/plz/controller/results/results_base.py]: Maybe delete this and call `is_finished`. 😛

File: services/controller/src/plz/controller/instances/aws/ec2_instance.py [hunk 4]
```diff
@@ -35,25 +35,33 @@ def __init__(self,
...
-        self.images.pull(snapshot_id)
-        self.delegate.run(command, snapshot_id, parameters, input_stream,
-                          docker_run_args)
+    def run_if_free(self,
+                    command: List[str],
+                    snapshot_id: str,
+                    parameters: Parameters,
+                    input_stream: Optional[io.BytesIO],
+                    docker_run_args: Dict[str, str],
+                    max_idle_seconds: int = 0) -> bool:
```
Comment (SamirTalwar) [inline services/controller/src/plz/controller/instances/aws/ec2_instance.py]: I think I'd prefer if this was still named `run` and raised a `ResourceConflictException` or something if it's not free. Too easy to forget to check a boolean.

File: services/controller/src/plz/controller/instances/aws/ec2_instance_group.py [hunk 5]
```diff
@@ -95,13 +97,16 @@ def instance_initialization_code(self) -> str:
         ])
         return self._instance_initialization_code
 
-    def acquire_instance(
+    def run_in_instance(
```
Comment (SamirTalwar) [inline services/controller/src/plz/controller/instances/aws/ec2_instance_group.py]: Bit weird to have a "release" without an "acquire". Is `release_instance` still used? If it's part of cleanup/harvesting, then let's rename it accordingly.

Comment (SamirTalwar) [inline services/controller/src/plz/controller/instances/aws/ec2_instance_group.py]: Bit weird to have a "release" without an "acquire". Is `release_instance` still used? If it's part of cleanup/harvesting, then let's rename it accordingly.

Comment (sergiogiro) [reply services/controller/src/plz/controller/instances/aws/ec2_instance_group.py]: We're gonna revisit this soon anyway...

**Pull request merged**